tests/templates/test_subroutines/test_grover.py

```diff
@@ -82,6 +85,76 @@ def decomposition_wires(wires):
     return wire_order
 
 
+def test_grover_diffusion_matrix_results():
+    """Test that the matrix gives the same result as when running the example in the documentation `here <https://pennylane.readthedocs.io/en/stable/code/api/pennylane.templates.subroutines.GroverOperator.html>`_"""
+    n_wires = 3
+    wires = list(range(n_wires))
+
+    def oracle():
+        qml.Hadamard(wires[-1])
+        qml.Toffoli(wires=wires)
+        qml.Hadamard(wires[-1])
+
+    dev = qml.device("default.qubit", wires=wires)
+
+    @qml.qnode(dev)
+    def GroverSearch(num_iterations=1):
+        for wire in wires:
+            qml.Hadamard(wire)
+
+        for _ in range(num_iterations):
+            oracle()
+            qml.templates.GroverOperator(wires=wires)
+        return qml.probs(wires)
+
+    # Get probabilities from example
+    probs_example = GroverSearch(num_iterations=1)
+
+    # Grover diffusion matrix
+    G_matrix = qml.templates.GroverOperator(wires=wires).matrix
+
+    oracle_matrix = np.identity(2 ** n_wires)
+    oracle_matrix[-1, -1] = -1
+
+    # state |0>
+    state_0 = [1, 0]
+    s1 = Hadamard.matrix @ state_0
+    # uniform superposition state
+    s = functools.reduce(np.kron, list(itertools.repeat(s1, n_wires)))
+
+    amplitudes = G_matrix @ oracle_matrix @ s
+    # Chek that the probabilities are the same
+    probs_matrix = amplitudes ** 2
+
+    assert np.allclose(probs_example, probs_matrix)
+
+
+def test_grover_diffusion_matrix():
+    """Test that the Grover diffusion matrix is the same as when constructed in a different way"""
+    n_wires = 4  # test with 4 wires
+    wires = list(range(n_wires))
+    G_matrix = qml.templates.GroverOperator(wires=wires).matrix
+
+    # Create Grover diffusion matrix G in alternative way
+    oplist = list(itertools.repeat(Hadamard.matrix, n_wires - 1))
+    oplist.append(PauliZ.matrix)
+
+    ctrl_str = "0" * (n_wires - 1)
+    CX = MultiControlledX(
+        control_values=ctrl_str,
+        control_wires=wires[:-1],
+        wires=wires[-1],
+        work_wires=None,
+    ).matrix
+
+    M = functools.reduce(np.kron, oplist)
+    G = M @ CX @ M
+    print(G, G_matrix)
+
+    # There is just a difference in sign
+    assert np.allclose(np.abs(G_matrix), np.abs(G))
```

Comment:
How come there's a difference in sign? :thinking:

Reply:
Yeah that's a good question. I don't know, as I'm not familiar with the two different representations and how the circuit representation is derived. But I guess it maybe doesn't matter much since it's just a global phase?

Reply:
In case this is unsatisfactory, I now made a change where I chose an oracle to test with, and then compare the resulting probabilities instead of the matrices. Then this pesky problem doesn't appear...